### ISSUE TITLE
Request addition of double support in ChanTraits.h

### ISSUE THREAD

**Issue by IntellectualKitty** [cinder/Cinder#1813]
While attempting to perform high-frequency pixel sampling to convert 2D spherical texture maps to cube maps, I was planning to use `ColorAT<double>` to sum a number of lower-precision color values and then take the average later.   However, the `CHANTRAIT<T>` templates do not provide support for doubles.

While I can, of course, work around this by manually defining R/G/B/A double values for summing (and I'll have to do this in the meantime), I don't feel like this is the cleanest approach.

**Comment by paulhoux** [comment]
Out of curiosity, what do you hope to win by using doubles? When adding less than a few thousand colors using floats, each with a value between 0 and 1, the rounding error never exceeds 1/255. So you might as well use floats, instead of doubles.

**Comment by IntellectualKitty** [comment]
While I am not fond of floats because of their relatively low precision – I think of them as the shorts of the floating-point world – the real issue is speed.  I cannot speak about all 64-bit architectures, but on my 2011 MacBook Pro at least, doubles are significantly faster than floats.

The issue is basically that the CPU converts 32-bit (and shorter) types to full 64-bit types before performing computations, sign-extending integer types and performing full conversions for floating-point types.  So, for each computation that isn't performed on values already in registers (or that remain in registers), it's basically a five-step process: (1) load, (2) convert to 64-bit, (3) perform the computation, (4) convert back to 32-bit (or lower), and (5) store.  For types that are already in 64-bit format, steps 2 and 4 can be omitted.

I've been programming for nearly 35 years, and I always thought of doubles as slow-moving dinosaurs compared to floats.  So, I was shocked when I tested this on my laptop a year or two back and found that for basic computations doubles were nearly twice as fast as floats.  The main exception was trigonometric functions like `sin(double x)` and `cos(double x)` which ended up being a bear when using doubles. 

In my early tests converting 2D spherical texture maps to texture cube maps (for rendering planets), I found that I had to sample each spherical texture pixel using a minimum 5x5 sub-pixel grid to hit every texture cube map pixel at least once.  I've only tested it with relatively small maps so far – 2048x1024 instead of 8192x4096 or even 16384x8192 – so I don't know if that 5x5 minimum sampling ratio will change, but the process is computationally intensive (e.g., a total of 52,428,800 samples for the 2048x1024 @ 5x5 case) and _it takes what feels like 20-30 seconds every single time_ (I haven't actually timed it, it just _feels that bad_).  So, if I could halve that by using doubles instead of floats, I'd be a happy man!

(For those who are interested, the code is working but it is not yet optimized.  The color lookup into the spherical map is performed just once – however I am still using `ColorAT<T> SurfaceT<T>::getPixel( ivec2 pos ) const` instead of the built-in `SurfaceT<T>::ConstIter` – and then distributed to the temporary cube sampling faces 25 times (or the square of whatever the `samplingRate` is parameterized to be): since I need to keep track of the number of times that a cube map pixel has been hit as well its color, I'm using a custom class `Sample` stored in six vectors, one per cube face.  I opted to use `ColorAf` as @paulhoux suggested, but I will probably switch to unencapsulated doubles instead.  Finally, I haven't yet pre-computed the trigonometric values needed to derive the cubic texture coordinate, but that will be my first step when I start optimizing the code.)

**Comment by MikeGitb** [comment]
I haven't benchmarked myself, but I find this very interesting. I'd only expect a slowdown when part of your variables are doubles or integral types, because then the standard mandates that the operands be promoted to doubles. However, if you e.g. multiply two floats and store the result in a float variable I would have expected that no conversion goes on. Am I wrong?
Have you checked if enabeling fastmath and/or avx helps?

**Comment by IntellectualKitty** [comment]
@MikeGitb I'm only using the default settings.

Also, this has nothing to do with the standard but rather is about simplifying the design and implementation of 64-bit chip sets.  Basically, the CPU internally promotes floats to doubles before values enter the 64-bit only computation pathway and then converts them back to floats afterward before storing them out to memory.  This saves them from having to implement a 32-bit computation pathway on the CPU as well which makes the chip smaller and less expensive.  Basically, they do away with native 32-bit computation pathways by using only mandatory operations since they really only have to implement (1) 64-bit computation pathways and (2) 64 <-> 32 conversions.  So, 32-bit computation is slower as a result.

In my current code, I can report that when I copy the `ColorAf` value returned from `ColorAT<T> SurfaceT<T>::getPixel( ivec2 pos ) const` to an array of doubles and also use an array of doubles in my `Sample` class, I get a 12% speedup versus using `ColorAf` in both places.  This 12% speed increase occurs despite the fact that my storage requirements have nearly doubled meaning that nearly twice as much data is moving through the cache.  Since my storage requirements are significant (e.g., 6 512x512 cube face samples are used when converting a 2048x1024 spherical map) and only simple addition is being performed, you'd see a significantly greater speed increase if more mathematical operations were performed or if the computations were being performed inline rather than constantly transmitting to and from a large chunk of memory.

**Comment by MikeGitb** [comment]
@IntellectualKitty: I have no reason to doubt your results but I find your explanation hard to believe. Could you point me to some documentation, that shows arithmetic instructions on 32 bit floats are slower than on 64 bit on a modern processor?

**Comment by IntellectualKitty** [comment]
@MikeGitb I'm afraid that I do not have whatever documentation that lead me down this path a year or two back (I think it was more than two years ago really).  Also, I'm not finding any substantive documentation when I Google (I'm not a notably proficient web searcher).  But I do come across a lot of informal results that indicate surprise that doubles are as fast as floats as well as people who've had the same experience that I have that they are faster (as well as results that say that floats are invariably faster).

For instance, I found the following comment that is similar to but more explicit than the explanation I gave at https://social.msdn.microsoft.com/Forums/vstudio/en-US/ed60d53f-472e-4a2a-aeb7-31ab40853e20/float-vs-double-performance?forum=csharpgeneral :

> Because the x86 CPU only has instructions for processing 80-bit floating point numbers.  So, both float and double get upconverted to 80-bits and then downconverted to the expected type (32-bit in the case of float, 64-bit in the case of double).

Here's a link to another random article (http://articles.emptycrate.com/2012/02/11/double_vs_float_which_is_better.html) where the author finds that long doubles are faster than doubles and doubles are faster than floats.  He includes some test code as part of his argument that you might want to try out on your system (I haven't run it).

You might want to check out https://software.intel.com/en-us/forums .  I saw a number of discussions from there (like this one at https://software.intel.com/en-us/forums/intel-isa-extensions/topic/393951) when I was searching.  There are a lot of varied opinions from people who all seem to be FPU knowledgable.  

Again, apart from little articles like this and people having similar discussions on other forums, I wasn't able to find anything significant as far as documentation.  I apologize for that.  Again, I am not a good web searcher, so perhaps I'm looking in the wrong place.

Would you mind doing some testing on your own system?  The results on mine have been very consistent and I'd like see what you (and others) are getting.  I'm on a late 2011 13-inch MacBook Pro with a 2.4 GHz Intel Core i5 processor and 4 GB of 1333 MHz DDR3 RAM running macOS 10.12.4.

**Comment by IntellectualKitty** [comment]
Closing this issue per PR #1818.

**Comment by IntellectualKitty** [comment]
After doing further reading, there still doesn't seem to be any consistent answer.  The only conclusion that I can come to is that it seems to be platform dependent.  So, I would advise testing on the systems that you intend to target.

**Comment by MikeGitb** [comment]
@IntellectualKitty: Thank you very much, that gives me something to do over the weekend ;)

**Comment by MikeGitb** [comment]
Regarding the benchmark: Anything in particular you would like me to run? Otherwise I can just put together some arbitrary micro-benchmark and put up the code together with the results.

Sorry for leading the thread astray a bit.

**Comment by MikeGitb** [comment]
@IntellectualKitty: Don't know, if this is still important to you, but I made a few basic tests on my ivy bridge machine with visual studio. For pure arithmetic I indeed notice a very slight performance degradation when using floats compared to doubles (well below 10% though). When doing things like matrix multiplication or transforming data stored in an array (data fits in l2 cache) floats where faster than doubles (~10 to 20%). I didn't try any explicit simd code though.

 A very important pitfall I noticed though was that this

    float foo(float a) {
         return static_cast<float>(a*2.0); //static cast to emphasis that a conversion is going on here
     }

was ~20% slower than
  
    return a*2.0f;

So when dealing with floats at all, it seems to be important to not accidentially promote the calculations to double.
